src/zope/tales/tests/test_expressions.py

```diff
@@ -35,95 +31,156 @@ class Data(object):
-              ErrorGenerator = ErrorGenerator(),
-              )
+            vars=dict(
+                x=d,
+                y=Data(z=3),
+                b='boot',
+                B=2,
+                adapterTest=at,
+                dynamic='z',
+                eightBits=u'déjà vu'.encode('utf-8'),
+                ErrorGenerator=ErrorGenerator(),
+                callable=Callable(),
+                old_callable_class=OldStyleCallable,
             )
+        )
 
         self.engine = Engine
 
+        self.py3BrokenEightBits = "a b'd\\xc3\\xa9j\\xc3\\xa0 vu'"
+
+
+    def _compiled_expr(self, expr):
+        return self.engine.compile(expr) if isinstance(expr, str) else expr
+
+    def _check_evals_to(self, expr, result):
+        expr = self._compiled_expr(expr)
+        self.assertEqual(expr(self.context), result)
+        return expr
+
+    def _check_evals_to_instance(self, expr, result_kind):
+        expr = self._compiled_expr(expr)
+        self.assertIsInstance(expr(self.context), result_kind)
+        return expr
+
+    def _check_raises_compiler_error(self, expr_str, regex=None):
+        from zope.tales.tales import CompilerError
+        meth = self.assertRaises if regex is None else self.assertRaisesRegexp
+        args = (regex,) if regex is not None else ()
+        with meth(CompilerError, *args) as exc:
+            self.engine.compile(expr_str)
+        return exc.exception
+
+    def _check_subexpr_raises_compiler_error(self, expr, regexp):
+        from zope.tales.expressions import SubPathExpr
+        from zope.tales.tales import CompilerError
+        with self.assertRaisesRegexp(CompilerError,
+                                     regexp):
+            SubPathExpr(expr, None, self.engine)
 
-class ExpressionTests(ExpressionTestBase):
+
+class TestParsedExpressions(ExpressionTestBase):
+    # Whitebox tests of expressions that have been compiled by the engine
 
     def testSimple(self):
         expr = self.engine.compile('x')
-        context=self.context
-        self.assertEqual(expr(context), context.vars['x'])
+        self._check_evals_to(expr, self.context.vars['x'])
 
     def testPath(self):
         expr = self.engine.compile('x/y')
-        context=self.context
-        self.assertEqual(expr(context), context.vars['x'].y)
+        self._check_evals_to(expr, self.context.vars['x'].y)
+        self.assertEqual("standard expression ('x/y')", str(expr))
+        self.assertEqual("<PathExpr standard:'x/y'>", repr(expr))
+
+    def test_path_nocall_call(self):
+        self._check_evals_to('callable', 42)
+        self._check_evals_to('nocall:callable', self.context.vars['callable'])
+
+        self._check_evals_to_instance('old_callable_class', OldStyleCallable)
+        self._check_evals_to('nocall:old_callable_class', OldStyleCallable)
+
+    def test_path_exists(self):
+        self._check_evals_to('exists:x', 1)
+        self._check_evals_to('exists:' + 'i' + str(id(self)), 0)
```

Comment:
Does `exists:` really return `1`/`0` instead of `True`/`False`?

Reply:
It [sure does](https://github.com/zopefoundation/zope.tales/blob/master/src/zope/tales/expressions.py#L175). 

Only reason for sure I could figure there was that 0 and 1 are bytecode constants whereas on Python 2 True and False require a LOAD_GLOBAL bytecode. So it's probably a speed hack.

Reply:
Wheee!

Or maybe it was written back when Python 2.1 was still used in Zopelandia.

Reply:
Heh. Oh, right. [Lest we forget](http://python-history.blogspot.com/2013/11/the-history-of-bool-true-and-false.html).